executor/include/executor/best_fit.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2014-, Open Perception, Inc.
+ *  Author(s): Shrijit Singh <shrijitsingh99@gmail.com>
+ *
+ */
+
+#pragma once
+
+#include <executor/type_trait.h>
+
+#include <executor/default/cuda_executor.hpp>
+#include <executor/default/inline_executor.hpp>
+#include <executor/default/omp_executor.hpp>
+#include <executor/default/sse_executor.hpp>
+
+namespace executor {
+
+static const auto best_fit_executors =
+    std::make_tuple(executor::omp_executor<>{}, executor::sse_executor<>{},
+                    executor::inline_executor<>{});
+
+struct executor_runtime_checks {
+  template <typename Executor, typename executor::InstanceOfAny<
+                                   Executor, executor::inline_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOfAny<
+                                   Executor, executor::sse_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOfAny<
+                                   Executor, executor::omp_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOfAny<
+                                   Executor, executor::cuda_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+};
+
+namespace detail {
+
+template <typename Function, typename Executor, typename = void>
+bool execute(Function&& f, Executor& exec) {
+  return false;
+}
+
+template <typename Function, template <typename...> class Executor, typename... Properties,
+          typename std::enable_if<executor::is_executor_available_v<Executor>,
+                                  int>::type = 0>
+bool execute(Function&& f, Executor<Properties...>& exec) {
+  f(exec);
+  return true;
+}
```

Comment:
I'm surprised you're not getting ambiguous function call errors here. I'm not seeing you disabling the first overload in case the executor is available.

Reply:
The second one is more specialized than the first, so if the second is valid the compiler should choose that right?